HACKING.md

```diff
@@ -79,96 +79,197 @@ The bigger PR will likely take some time to get reviewed, and discussed, while s
+* Fold new code at 78 columns.
+    * Do not break string literals.
+      Prefer having strings on a single line, in order to improve search-ability(grep).
+      If they do not fit on a single line, try breaking on punctuation.
+      You can surpass the 78 column limit if necessary.
+* Use single spaces around operators, but not before semicolon:
   ```c
-  Log(LOG_LEVEL_INFO, "Some error occurred while reading installed "
-      "packages cache.");
+  int a = 2 + 3;
+  int b = 2 * (a - 2);
+  if (a > b)
   ```
-  Good:
+* Place pointer star to the right, next to the name it belongs to:
+  ```c
+  int *a;
+  char *b, *c;
+  const char *const d;
+  ```
+* Curly brackets on separate lines.
+    * Except in a do-while loop, to avoid confusion with a while loop:
+      ```c
+      do
+      {
+          // ...
+      } while (condition);
+      ```
+    * Except at the end of a struct type definition:
+      ```c
+      typedef struct _Name
+      {
+          // ...
+      } Name;
+      ```
+* Type casts should be separated with one space from the variable:
+  ```c
+  (struct sockaddr *) &myaddr
+  ```
+
+### Readability / Maintainability
+
+Keep in mind that code should be readable by non C experts.
+If you are a Guru, try to restrain yourself, only do magic when absolutely necessary.
+
+* Avoid code duplication.
+  Make a variable or a function instead of copy-pasting the same patterns.
+    * Helper functions which are only used in the same file should be declared `static`.
+    * If you want to test the `static` functions, you can `#include` the `.c` file in the unit test.
+* Avoid abbreviations.
+* Welcome to "modern" C (C99).
+    * Declare variables where they are needed, not at the beginning of a function / block.
+    * Declare index inside for loop:
+      ```c
+      for (int i = 0; i < n; ++i)
+      ```
+* Variable names:
+    * Describe what the variable is.
+    * Are English words with underscores: `string_index`
+    * Boolean variable names should make an `if`/`while` sensible: `if (more_data)` / `while (more_data)`
+* Function names:
+    * Are `CamelCase` (with first letter capital).
+    * Describe what the function does.
+* Namespaces:
+    * Don't exist in C.
+    * But we can pretend, using underscores in function names:
+        * `StrList_BinarySearch()`
+* Structs:
+    * Names are `CamelCase` (with first letter capital).
+    * Use the common typedef pattern:
+      ```c
+      typedef struct _Point
+      {
+          float x;
+          float y;
+      } Point;
+      ```
+      Don't add a trailing `_t` or similar for typedefs.
+    * Structs and functions which modify them are similar to classes and methods in other languages.
+        * Method names should start with the struct name.
+        * The first parameter should be the "self" pointer:
+          ```c
+          void SeqAppend(Seq *seq, void *item);
+          ```
+* Document using Doxygen (within reason), preferably in the `.c` files,
```

Comment:
Looks like the markup is wrong here (it's rendered in italics instead of as a bullet point). Also, I don't really agree with this. I hate to open `.c` files only to read what a particular function does. I wish we had all the doc comments in the header files.

Reply:
I agree with having documentation in the header files. Long documentation entries can clutter the code, and makes navigating the file slower.

The exception is for `static` functions obviously.

Reply:
This renders ok on github. Just a problem with their syntax highlighter.

Let's discuss these things in a new PR.